### ISSUE TITLE
Services that write to disk should detect disk pressure and clean up

### ISSUE THREAD

**Issue by nicksnyder** [sourcegraph/sourcegraph#627]
Specifically services that write to SSD locations which are a shared resource on the node (e.g. xlang-go-bg).

For example, if a service detects the location it is writing to is near capacity (e.g. 90% used), then it should perform some cleanup actions.

**Comment by chrismwendt** [comment]
I propose that the cache directory be treated as volatile - any file you write to that directory can be deleted at any time. That will allow us to write a simple garbage collector process that wakes up every minute and deletes everything in the cache if it's above 90% usage. Does anyone know if this property would break any of our services?

For ROI calculations: manually responding to the opsgenie alerts for SSDDiskSpaceCritical takes ~30 minutes each time, and happens about once every two weeks.

**Comment by slimsag** [comment]
https://github.com/sourcegraph/sourcegraph/issues/1759 I would be careful about how this happens atomically, not saying we will run into this issue but just saying we could so it is one aspect to consider here.

**Comment by keegancsmith** [comment]
We already have this in two ways:
- pod-tmp-gc is a Daemonset which deletes the cache dirs for pods that are not running
- Our disk cache logic supports this, and our services use that logic. However, we are not using it correctly in some cases.

pod-tmp-gc would be a good place to do this extra monitoring and implement the eviction when our application code is not behaving correctly. Note: it may not work without the application behaving properly, since deleting a file in linux only reclaims the disk space once all file handles to it are closed.

**Comment by nicksnyder** [comment]
What disk cache logic are you referring to in the second bullet?

**Comment by keegancsmith** [comment]
I spoke too soon, we actually do it differently now (in searcher). We can build this bit of logic into lang-go/etc. https://github.com/sourcegraph/sourcegraph/blob/master/pkg/diskcache/cache.go#L194

**Comment by nicksnyder** [comment]
We have non-Go binaries too (e.g. lang-typescript). Do you think we should add similar logic to those services or enhance the cleanup logic in a global garbage collector (like pod-top-gc) similar to what Chris proposed? To me, the latter seems like an easier way to assert the behavior we want.

**Comment by keegancsmith** [comment]
Yeah the latter sounds good. We must remember to include the binary in our server image.

**Comment by slimsag** [comment]
Does pod-tmp-gc get ran inside all of our existing containers? Or is it something that we must deploy separately?

**Comment by nicksnyder** [comment]
Yes, it is deployed as a DaemonSet so it runs one process per node.
https://github.com/sourcegraph/deploy-sourcegraph/tree/master/configure/ssd

**Comment by slimsag** [comment]
I see, I think it will be hard / complex to translate this into a pure-Docker deployment. What is the reason we don't bake this logic into our container images directly?

**Comment by nicksnyder** [comment]
Given that pod-tmp-gc already exists, it seems like we already have a problem with the Docker deployment if this isn't already handled?

The problem statement is: 
> In the presence of multiple services which use different tech stacks (e.g. Go, TypeScript), how do we ensure that a shared disk with finite space doesn't fill up?

Requiring each service to be a good citizen is one direction we could take (and appears to be the direction he have taken up until this point), but judging by the alerts it doesn't seem like this approach is working, and it seems like a pain to maintain this behavior across different services and tech stacks.

It is much easier to verify that the system behaves as desired (i.e. the disk will never fill up), if we have a single agent that is responsible for doing so. Also, given that the resource we are managing is per node, it seems natural to have a service whose lifecycle matches the lifecycle of the resource that it manages.

**Comment by ggilmore** [comment]
> I see, I think it will be hard / complex to translate this into a pure-Docker deployment. What is the reason we don't bake this logic into our container images directly?

The containers themselves can't run this logic because they'll be killed before any sort of clean up routine they invoke has a chance to run (more simply, the thing that's being killed can't be expected to clean up after itself). 

If you want to avoid a separate pod-tmp-gc deployment, you could change the logic to have all containers running on a node run a modified version of `pod-tmp-gc` that is able to clean up after _other_ pods running on the same node. (The same atomicity concerns would stil apply here). 

You could also use a [pre-stop lifecycle hook](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/#container-hooks) to run this clean up logic before the container exits, but I don't know if that would be possible to translate to non-K8s environments.

**Comment by slimsag** [comment]
Also, do we not have the same problems with disk space filling up and preventing the container from starting in the case of Sourcegraph server? I would rephrase our problem statement here to:

> In the presence of multiple services which use different tech stacks (e.g. Go, TypeScript) and are deployed in different ways (Kubernetes, single-container Docker installation, multi-container Docker installation), how do we ensure that a shared disk with finite space doesn't fill up?

**Comment by ggilmore** [comment]
`pod-tmp-gc` currently works by:

1. Mounting the SSD which is shared amongst all pods that need a persistent cache running on a node

1. Querying the K8s api to get the set of all running pods on the current node

1. Iterating through all the top-level folders on the disk to see which ones are stale (i.e. left over from pods that have stopped)

1. Deleting all the stale folders on the disk

> Am I right that pod-tmp-gc needs to run once per persistent /tmp volume on the host node? i.e. once per service as I indicated above?

No, `pod-tmp-gc` looks over an entire physical disk (which could be shared between multiple services). 

> Why can't we have pod-tmp-gc do cleanup on container start?

> Why can't we have pod-tmp-gc run in the background in containers ...

`pod-tmp-gc` needs to know what pods are currently running so it can identify folders that are safe to delete. It currently gets this info from the K8s api. If `pod-tmp-gc` was baked into each pod, each pod would need additional k8s permissions so that it can query this information (and this wouldn't be portable outside of K8s). This would also imply that each pod would need to mount the _entire_ disk, which seems heavy-handed / dangerous.

> ... guarantee a certain amount of free cache disk space, to ensure the container can always start in the future?

"guarantee a certain amount" sounds like `pod-tmp-gc` would need to know some semantics about each application to know what folders are safe to delete.

**Comment by keegancsmith** [comment]
If I am not mistaken, the pure docker deployment mounts a stable and unique volume per replica. So it doesn't require pod-tmp-gc. However, it does require a service to cleanup since the volume "follows" the service around.

ie the difference between your pure docker deployment and kubernetes is that in kubernetes pods are ephemeral, but the (named) containers in docker in docker are persistent.

**Comment by chrismwendt** [comment]
@slimsag I believe so, that's my understanding of the problem space as well. We could add functionality to pod-tmp-gc so that pod-tmp-gc could **also** prevent problem 2 (unbounded growth of `/tmp`).

**Comment by slimsag** [comment]
@ggilmore https://sourcegraph.sgdev.org/search?q=repo:%5Egithub%5C.com/sourcegraph/deploy-sourcegraph-dot-com%24+ssd0

These services are all sharing the same SSD. This implies that either ALL of these services are running on the same host node, or that the SSD is actually something like NFS which means we will suffer from network latency. Is it the former (we are deploying all of those services onto the same host node)?

**Comment by keegancsmith** [comment]
The SSD is local to the node. We configure GCE to give us VMs with local SSDs. Kubernetes is then configured with a host path to that directory. See the funky magic we do with mounting the cache dir, where the root path is `/path/to/local/ssd/pod-tmp-gc/${POD_NAME}`

**Comment by slimsag** [comment]
Thanks, that makes sense. So then this issue is just about solving the second issue (the first is solved by pod-tmp-gc already):

> 2. In ALL deployments, cleaning up existing /tmp directories by e.g. ensuring they do not grow unbounded, etc.

And this is an issue that we will have to solve not just in Kubernetes deployments, but also in Server (single-container) and pure-Docker (multi-container) deployments.

i.e. we will also need some process to ensure /tmp does not grow unbounded in Server (single-container) deployments, or else it will eventually eat all of the disk space and crash

**Comment by slimsag** [comment]
Since I fear I may have derailed the conversation a bit by asking my questions, what are the problems / objections with my proposal to solve issue #2 above:

- We build a Go binary which can be ran in the background of all existing containers to periodically clean up the active pod/service's /tmp directory, e.g. by deleting it at some interval

**Comment by ggilmore** [comment]
We'd need to know that the temp directory is safe to delete somehow right? (e.g. we don't want some operation to fail because a file was garbage collected in the middle of an operation)

**Comment by slimsag** [comment]
Me / Geoffrey / Nick talked over a call, conclusions:

- We agree a separate Go binary that runs in the background of our existing Docker containers is the right path forward.
- It will _enforce_ that we do not allow /tmp to grow unbounded, but services themselves should still be aware of cleaning up after themselves. i.e. this service acts as a 'disk usage limit' rather than something that cleans up after services _for them_, it doesn't relieve services of their responsibility of ensuring disk usage does not grow unbounded.
- By doing this, we shift bugs from "I am a Sourcegraph user and have to constantly clean up /tmp" to "I am a Sourcegraph dev and we have a process that cleans up /tmp by deleting it when it grows too large, my service should be resilient to that fact"

**Comment by keegancsmith** [comment]
Our most common writer to /tmp is archives, and the code which does that is resillient to /tmp being rmed (it writes to a tmp file and renames into place once complete)

**Comment by nicksnyder** [comment]
Most recent instance of this alert https://sourcegraph.slack.com/archives/C08JA8Q1H/p1549471151205000

The biggest space consumers were sourcegraph-frontend and lang-go.

```
/home/node/cache # du -d 1 -h
1.9G    ./lang-typescript-785d6f85fb-985gv
169.1G    ./lang-go-fccc75799-2wpb9
450.6M    ./lang-typescript-785d6f85fb-mqxzs
2.2G    ./lang-typescript-785d6f85fb-8kvww
4.0K    ./pod-tmp-gc
34.0G    ./searcher-6bfc4bf5f9-hhggc
130.1G    ./sourcegraph-frontend-69b9d4c699-67ktc
337.7G    .
```

**Comment by tsenart** [comment]
@keegancsmith and I handled another instance of this alert. This is a ticking time-bomb for customers upgrading to 3.0.0, albeit a slower one than our sourcegraph.com deployment.

<img width="1874" alt="screenshot 2019-02-19 at 15 09 33" src="https://user-images.githubusercontent.com/67471/53020974-6f43ee00-3458-11e9-899e-785dba773b81.png">

**Comment by tsenart** [comment]
By the way, the issue this time around was with the frontend.
